usr/sbin/p11sak/p11sak.c

```diff
@@ -113,20 +114,57 @@ static CK_RV get_pin(char **pin, size_t *pinlen) {
     }
 
     /* strip the carriage return since not part of pin. */
-    buff[nread - 1] = '\0';
-    memcpy(*pin, buff, nread);
+    user_input[nread - 1] = '\0';
+    memcpy(*pin, user_input, nread);
     /* don't include the terminating null in the pinlen */
     *pinlen = nread - 1;
 
-    done: if (buff)
-        free(buff);
+    done: if (user_input)
+        free(user_input);
+
+    return rc;
+}
+
+static CK_RV get_keyboard_input(char **user_input)
+{
+    int nread;
+    char *keyboard_input = NULL;
+    size_t buflen;
+    CK_RV rc = 0;
+
+    /* read the keyboard_input
+     * Note: getline will allocate memory for keyboard_input. free it when done.
+     */
+    nread = getline(&keyboard_input, &buflen, stdin);
+    if (nread == -1) {
+        rc = -1;
+        goto done;
+    }
+
+    /* Allocate keyboard_input.
```

Comment:
See above.
If you really need to duplicate the buffer, why not simply use strdup() ?

The buffer allocated by getline is supposed to be zero terminated, so stdup should just work fine.

Reply:
If your intention was to only modify the user_input arg in case of successful exit, you could just set user_input = keyboard_input in that case, instead of allocating and copying to a new buffer. Caller is responsible to free user_input any way.

Reply:
Simplified user input into confirm_destroy()